source/compute_intersections.cc

```diff
@@ -0,0 +1,375 @@
+  std::vector<std::tuple<
+    typename dealii::Triangulation<dim0, spacedim>::active_cell_iterator,
+    typename dealii::Triangulation<dim1, spacedim>::active_cell_iterator,
+    dealii::Quadrature<spacedim>>>
+  compute_intersection(const GridTools::Cache<dim0, spacedim> &space_cache,
+                       const GridTools::Cache<dim1, spacedim> &immersed_cache,
+                       const unsigned int                      degree)
+  {
+    Assert(DEAL_II_WITH_CGAL,
+           ExcMessage("This function needs CGAL to be installed, "
+                      "but cmake could not find it."));
+
+    Assert(degree >= 1, ExcMessage("degree cannot be 0"));
+
+    std::set<typename Triangulation<dim1, spacedim>::active_cell_iterator>
+                intersected_cells; // avoid duplicates
+    const auto &space_tree =
+      space_cache.get_locally_owned_cell_bounding_boxes_rtree();
+    const auto &immersed_tree =
+      immersed_cache.get_locally_owned_cell_bounding_boxes_rtree();
+
+
+    namespace bgi = boost::geometry::index;
+    // Whenever the BB space_cell intersects the BB of an embedded cell, store
+    // the space_cell in the set of intersected_cells
+    for (const auto &[immersed_box, immersed_cell] : immersed_tree)
+      {
+        for (const auto &[space_box, space_cell] :
+             space_tree | bgi::adaptors::queried(bgi::intersects(immersed_box)))
+          {
+            intersected_cells.insert(space_cell);
+          }
+      } // found intersected cells
+
+
+    // references to triangulations' info (cp cstrs marked as delete)
+    const auto &mapping0 = space_cache.get_mapping();
+    const auto &mapping1 = immersed_cache.get_mapping();
+
+    std::vector<
+      std::tuple<typename Triangulation<dim0, spacedim>::active_cell_iterator,
+                 typename Triangulation<dim1, spacedim>::active_cell_iterator,
+                 Quadrature<spacedim>>>
+      cells_with_quads;
+
+    for (const auto &space_cell : intersected_cells)
+      { // loop over interseced space_cells
+        for (const auto &[immersed_box, immersed_cell] : immersed_tree)
+          {
+            typename Triangulation<dim0, spacedim>::active_cell_iterator
+              space_cell_t(space_cell);
+
+            typename Triangulation<dim1, spacedim>::active_cell_iterator
+              immersed_cell_t(immersed_cell);
+
+            const auto test_intersection =
+              compute_intersection<dim0, dim1, spacedim>(
+                space_cell_t, immersed_cell_t, degree, mapping0, mapping1);
+
+            if (test_intersection.get_points().size() != 0)
+              {
+                cells_with_quads.push_back(std::make_tuple(space_cell_t,
+                                                           immersed_cell_t,
+                                                           test_intersection));
+              }
+          }
+      }
+
+    return cells_with_quads;
+  }
+
+} // namespace dealii::NonMatching
+
```

Comment:
Add an `#else`  and `#endif` here, where you have the default implementations that simply throws exceptions saying that you need CGAL, and then instantiate all of the functions irrespective of wether CGAL exists or not. This makes sure you don't get errors in user codes at link time, but just at run time.